File: services/orchestrator/app/contracts/a2ui.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field


class Binding(BaseModel):
    prop: str
    path: str


class SurfaceComponent(BaseModel):
    id: str
    component: str
    props: Optional[Dict[str, Any]] = None
    bindings: Optional[List[Binding]] = None


class DataUpdate(BaseModel):
    path: str
    op: Literal["replace", "append", "merge"]
    value: Any


class EnvelopeBase(BaseModel):
    request_id: str = Field(alias="requestId")
    timestamp: datetime

    model_config = ConfigDict(populate_by_name=True)


class SurfaceUpdateEnvelope(EnvelopeBase):
    type: Literal["surfaceUpdate"]
    components: List[SurfaceComponent]


class DataModelUpdateEnvelope(EnvelopeBase):
    type: Literal["dataModelUpdate"]
    updates: List[DataUpdate]


class BeginRenderingEnvelope(EnvelopeBase):
    type: Literal["beginRendering"]
    surface_id: str = Field(alias="surfaceId")


A2UIEnvelope = Union[SurfaceUpdateEnvelope, DataModelUpdateEnvelope, BeginRenderingEnvelope]

File: src/lib/contracts/a2ui.ts
import { z } from "zod";

const bindingSchema = z.object({
  prop: z.string().min(1),
  path: z.string().min(1)
});

const componentSchema = z.object({
  id: z.string().min(1),
  component: z.string().min(1),
  props: z.record(z.unknown()).optional(),
  bindings: z.array(bindingSchema).optional()
});

const dataUpdateSchema = z.object({
  path: z.string().min(1),
  op: z.enum(["replace", "append", "merge"]),
  value: z.unknown()
});

const envelopeBaseSchema = z.object({
  requestId: z.string().min(1),
  // JSON Schema `date-time` allows timezone offsets (e.g. +00:00).
  timestamp: z.string().datetime({ offset: true })
});

const surfaceUpdateSchema = envelopeBaseSchema.extend({
  type: z.literal("surfaceUpdate"),
  components: z.array(componentSchema).min(1)
});

const dataModelUpdateSchema = envelopeBaseSchema.extend({
  type: z.literal("dataModelUpdate"),
  updates: z.array(dataUpdateSchema).min(1)
});

const beginRenderingSchema = envelopeBaseSchema.extend({
  type: z.literal("beginRendering"),
  surfaceId: z.string().min(1)
});

export const a2uiEnvelopeSchema = z.discriminatedUnion("type", [
  surfaceUpdateSchema,
  dataModelUpdateSchema,
  beginRenderingSchema
]);

export type A2UIEnvelope = z.infer<typeof a2uiEnvelopeSchema>;
